sys/net/application_layer/gcoap/forward_proxy/forward_proxy.c

```diff
@@ -0,0 +1,232 @@
+
+    /* parse destination address */
+    if (ipv6_addr_from_str(&addr, addr_str) == NULL) {
+        return false;
+    }
+    if ((remote->netif == SOCK_ADDR_ANY_NETIF) && ipv6_addr_is_link_local(&addr)) {
+        return false;
+    }
+    memcpy(&remote->addr.ipv6[0], &addr.u8[0], sizeof(addr.u8));
+
+    /* copy port string into scratch for atoi */
+    memcpy(scratch, urip->port, urip->port_len);
+    scratch[urip->port_len] = '\0';
+
+    remote->port = atoi((char *) scratch);
+
+    if (remote->port == 0) {
+        return false;
+    }
+
+    return true;
+}
+
+static void _forward_resp_handler(const gcoap_request_memo_t *memo,
+                                  coap_pkt_t* pdu,
+                                  const sock_udp_ep_t *remote)
+{
+    (void) remote; /* this is the origin server */
+
+    /* forward the response packet as-is to the client */
+    gcoap_dispatch((uint8_t *)pdu->hdr,
+                   (pdu->payload - (uint8_t *)pdu->hdr + pdu->payload_len),
+                   (sock_udp_ep_t *)&memo->client_ep);
+}
+
+static int _gcoap_forward_proxy_via_coap(coap_pkt_t *client_pkt,
+                                         sock_udp_ep_t *client_ep,
+                                         uri_parser_result_t *urip)
+{
+    coap_pkt_t pkt;
+    sock_udp_ep_t origin_server_ep;
+
+    gcoap_request_memo_t *memo = NULL;
+
+    if (!_parse_endpoint(&origin_server_ep, urip)) {
+        return -EINVAL;
+    }
+
+    /* do not forward requests if they already exist, e.g., due to CON
+       and retransmissions. In the future, the proxy should set an
+       empty ACK message to stop the retransmissions of a client */
+    gcoap_find_req_memo(&memo, client_pkt, &origin_server_ep);
+    if (memo) {
+        DEBUG("gcoap_forward_proxy: request already exists, ignore!\n");
+        return 0;
+    }
+
+    unsigned token_len = coap_get_token_len(client_pkt);
+
+    coap_pkt_init(&pkt, proxy_req_buf,
+                  (CONFIG_GCOAP_PDU_BUF_SIZE - CONFIG_GCOAP_REQ_OPTIONS_BUF),
+                  sizeof(coap_hdr_t) + token_len);
+
+    pkt.hdr->ver_t_tkl = client_pkt->hdr->ver_t_tkl;
+    pkt.hdr->code = client_pkt->hdr->code;
+    pkt.hdr->id = client_pkt->hdr->id;
+
+    if (token_len) {
+        memcpy(pkt.token, client_pkt->token, token_len);
+    }
+
+    ssize_t res = coap_opt_add_chars(&pkt, COAP_OPT_URI_PATH,
+                                     urip->path, urip->path_len, '/');
+    if (res < 0) {
+        return -EINVAL;
+    }
+
+    if (urip->query) {
+        res = coap_opt_add_chars(&pkt, COAP_OPT_URI_PATH,
+                                 urip->query, urip->path_len, '&');
+        if (res < 0) {
+            return -EINVAL;
+        }
+    }
+
+    /* copy all options from client_pkt to pkt */
+    coap_optpos_t opt = {0, 0};
+    uint8_t *value;
+    for (int i = 0; i < client_pkt->options_len; i++) {
+        ssize_t optlen = coap_opt_get_next(client_pkt, &opt, &value, !i);
+        if (optlen >= 0) {
```

Comment:
Encountering a non-safe-to-forward option needs to result in an error response here.